storage/pebble/execution_data_tracker.go

```diff
@@ -0,0 +1,406 @@
+// - startHeight: The initial fulfilled height to be set if no previous fulfilled height is found.
+// - opts: Additional configuration options such as custom prune callbacks.
+//
+// No errors are expected during normal operation.
+func NewExecutionDataTracker(
+	logger zerolog.Logger,
+	path string,
+	startHeight uint64,
+	opts ...StorageOption,
+) (*ExecutionDataTracker, error) {
+	lg := logger.With().Str("module", "tracker_storage").Logger()
+	db, err := pebble.Open(path, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not open db: %w", err)
+	}
+
+	tracker := &ExecutionDataTracker{
+		db:            db,
+		pruneCallback: func(c cid.Cid) error { return nil },
+		logger:        lg,
+	}
+
+	for _, opt := range opts {
+		opt(tracker)
+	}
+
+	lg.Info().Msgf("initialize tracker with start height: %d", startHeight)
+
+	if err := tracker.init(startHeight); err != nil {
+		return nil, fmt.Errorf("failed to initialize tracker: %w", err)
+	}
+
+	lg.Info().Msgf("tracker initialized")
+
+	return tracker, nil
+}
+
+// TODO: move common logic into separate function to avoid duplication of code
+// init initializes the ExecutionDataTracker by setting the fulfilled and pruned heights.
+//
+// Parameters:
+// - startHeight: The initial fulfilled height to be set if no previous fulfilled height is found.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) init(startHeight uint64) error {
+	fulfilledHeight, fulfilledHeightErr := s.GetFulfilledHeight()
+	prunedHeight, prunedHeightErr := s.GetPrunedHeight()
+
+	if fulfilledHeightErr == nil && prunedHeightErr == nil {
+		if prunedHeight > fulfilledHeight {
+			return fmt.Errorf(
+				"inconsistency detected: pruned height (%d) is greater than fulfilled height (%d)",
+				prunedHeight,
+				fulfilledHeight,
+			)
+		}
+
+		s.logger.Info().Msgf("prune from height %v up to height %d", fulfilledHeight, prunedHeight)
+		// replay pruning in case it was interrupted during previous shutdown
+		if err := s.PruneUpToHeight(prunedHeight); err != nil {
+			return fmt.Errorf("failed to replay pruning: %w", err)
+		}
+		s.logger.Info().Msgf("finished pruning")
+	} else if errors.Is(fulfilledHeightErr, storage.ErrNotFound) && errors.Is(prunedHeightErr, storage.ErrNotFound) {
+		// db is empty, need to bootstrap it
+		if err := s.bootstrap(startHeight); err != nil {
+			return fmt.Errorf("failed to bootstrap storage: %w", err)
+		}
+	} else {
+		return multierror.Append(fulfilledHeightErr, prunedHeightErr).ErrorOrNil()
+	}
+
+	return nil
+}
+
+// bootstrap sets the initial fulfilled and pruned heights to startHeight in an empty database.
+//
+// Parameters:
+// - startHeight: The initial height to set for both fulfilled and pruned heights.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) bootstrap(startHeight uint64) error {
+	err := operation.InitTrackerFulfilledHeight(startHeight)(s.db)
```

Comment:
similarly, you should complete both of these in a single pebble batch

Reply:
[34e04ee](https://github.com/onflow/flow-go/pull/6277/commits/34e04eec57c1bfa33cf840f8b107cc507bcf6c24)